lib/protoflow-blocks/src/blocks/io/decode_hex.rs

```diff
@@ -0,0 +1,145 @@
+// This is free and unencumbered software released into the public domain.
+
+use crate::{
+    prelude::{Bytes, Vec},
+    IoBlocks, StdioConfig, StdioError, StdioSystem, System,
+};
+use protoflow_core::{Block, BlockResult, BlockRuntime, InputPort, OutputPort};
+use protoflow_derive::Block;
+use simple_mermaid::mermaid;
+
+/// A block that decodes a hexadecimal byte stream to byte.
+///
+/// # Block Diagram
+#[doc = mermaid!("../../../doc/io/decode_hex.mmd")]
+///
+/// # Sequence Diagram
+#[doc = mermaid!("../../../doc/io/decode_hex.seq.mmd" framed)]
+///
+/// # Examples
+///
+/// ## Using the block in a system
+///
+/// ```rust
+/// # use protoflow_blocks::*;
+/// # fn main() {
+/// System::build(|s| {
+///     let stdin = s.read_stdin();
+///     let hex_decoder = s.decode_hex();
+///     let stdout = s.write_stdout();
+///     s.connect(&stdin.output, &hex_decoder.input);
+///     s.connect(&hex_decoder.output, &stdout.input);
+/// });
+/// # }
+/// ```
+///
+/// ## Running the block via the CLI
+///
+/// ```console
+/// $ protoflow execute DecodeHex
+/// ```
+///
+#[derive(Block, Clone)]
+pub struct DecodeHex {
+    /// The input text stream.
+    #[input]
+    pub input: InputPort<Bytes>,
+
+    /// The output byte stream.
+    #[output]
+    pub output: OutputPort<Bytes>,
+}
+
+impl DecodeHex {
+    pub fn new(input: InputPort<Bytes>, output: OutputPort<Bytes>) -> Self {
+        Self { input, output }
+    }
+
+    pub fn with_system(system: &System) -> Self {
+        use crate::SystemBuilding;
+        Self::new(system.input(), system.output())
+    }
+}
+
+impl Block for DecodeHex {
+    fn execute(&mut self, runtime: &dyn BlockRuntime) -> BlockResult {
+        runtime.wait_for(&self.input)?;
+
+        while let Some(message) = self.input.recv()? {
+            let decoded = hex_to_bytes(&message);
+            self.output.send(&decoded)?;
+        }
+
+        Ok(())
+    }
+}
+fn hex_to_bytes(hex_message: &Bytes) -> Bytes {
+    let mut decoded = Vec::with_capacity(hex_message.len() / 2);
+
+    for chunk in hex_message.chunks_exact(2) {
+        let high = chunk[0];
+        let low = chunk[1];
+        decoded.push((hex_value(high) << 4) | hex_value(low));
+    }
+
+    Bytes::from(decoded)
+}
+
+#[inline(always)]
+fn hex_value(byte: u8) -> u8 {
+    match byte {
+        b'0'..=b'9' => byte - b'0',
+        b'a'..=b'f' => byte - b'a' + 10,
+        b'A'..=b'F' => byte - b'A' + 10,
+        _ => panic!("Invalid hex character"),
```

Comment:
Blocks aren't allowed to panic on invalid inputs. We'll need to figure out what a good error-handling strategy here would be, but for now could ignore invalid inputs while logging them with [tracing](https://crates.io/crates/tracing) if the `tracing` feature is enabled.

Reply:
done